
from dataclasses import dataclass
import typing
import typing_extensions

from flytekit.core.annotation import FlyteAnnotation

from latch.types.metadata import NextflowParameter
from latch.types.file import LatchFile
from latch.types.directory import LatchDir, LatchOutputDir

# Import these into your `__init__.py` file:
#
# from .parameters import generated_parameters

generated_parameters = {
    'input': NextflowParameter(
        type=str,
        default=None,
        section_title='Input/output options',
        description="Either paths or URLs to FASTQ/BAM data (must be surrounded with quotes). For paired end data, the path must use '{1,2}' notation to specify read pairs. Alternatively, a path to a TSV file (ending .tsv) containing file paths and sequencing/sample metadata. Allows for merging of multiple lanes/libraries/samples. Please see documentation for template.",
    ),
    'udg_type': NextflowParameter(
        type=typing.Optional[str],
        default='none',
        section_title=None,
        description="Specifies whether you have UDG treated libraries. Set to 'half' for partial treatment, or 'full' for UDG. If not set, libraries are assumed to have no UDG treatment ('none'). Not required for TSV input.",
    ),
    'single_stranded': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies that libraries are single stranded. Always affects MALTExtract but will be ignored by pileupCaller with TSV input. Not required for TSV input.',
    ),
    'single_end': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies that the input is single end reads. Not required for TSV input.',
    ),
    'colour_chemistry': NextflowParameter(
        type=typing.Optional[int],
        default=4,
        section_title=None,
        description='Specifies which Illumina sequencing chemistry was used. Used to inform whether to poly-G trim if turned on (see below). Not required for TSV input. Options: 2, 4.',
    ),
    'bam': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies that the input is in BAM format. Not required for TSV input.',
    ),
    'snpcapture_bed': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Input Data Additional Options',
        description='If library result of SNP capture, path to BED file containing SNPS positions on reference genome. SNP statistics are qualimap results directory only not MultiQC.',
    ),
    'run_convertinputbam': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turns on conversion of an input BAM file into FASTQ format to allow re-preprocessing (e.g. AdapterRemoval etc.).',
    ),
    'fasta': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Reference genome options',
        description="Path or URL to a FASTA reference file (required if not iGenome reference). File suffixes can be: '.fa', '.fn', '.fna', '.fasta'.",
    ),
    'genome': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Name of iGenomes reference (required if not FASTA reference). Requires argument `--igenomes_ignore false`, as iGenomes is ignored by default in nf-core/eager',
    ),
    'bwa_index': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Path to directory containing pre-made BWA indices (i.e. the directory before the files ending in '.amb' '.ann' '.bwt'. Do not include the files themselves. Most likely the same directory of the file provided with --fasta). If not supplied will be made for you.",
    ),
    'bt2_index': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Path to directory containing pre-made Bowtie2 indices (i.e. everything before the endings e.g. '.1.bt2', '.2.bt2', '.rev.1.bt2'. Most likely the same value as --fasta). If not supplied will be made for you.",
    ),
    'fasta_index': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Path to samtools FASTA index (typically ending in '.fai'). If not supplied will be made for you.",
    ),
    'seq_dict': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Path to picard sequence dictionary file (typically ending in '.dict'). If not supplied will be made for you.",
    ),
    'large_ref': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description="Specify to generate more recent '.csi' BAM indices. If your reference genome is larger than 3.5GB, this is recommended due to more efficient data handling with the '.csi' format over the older '.bai'.",
    ),
    'save_reference': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='If not already supplied by user, turns on saving of generated reference genome indices for later re-usage.',
    ),
    'outdir': NextflowParameter(
        type=typing.Optional[typing_extensions.Annotated[LatchDir, FlyteAnnotation({'output': True})]],
        default=None,
        section_title='Output options',
        description='The output directory where the results will be saved.',
    ),
    'email': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Generic options',
        description='Email address for completion summary.',
    ),
    'awsqueue': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Institutional config options',
        description='The AWSBatch JobQueue that needs to be set when running on AWSBatch',
    ),
    'awsregion': NextflowParameter(
        type=typing.Optional[str],
        default='eu-west-1',
        section_title=None,
        description='The AWS Region for your AWS Batch job to run on',
    ),
    'awscli': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to the AWS CLI tool',
    ),
    'skip_fastqc': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Skip steps',
        description=None,
    ),
    'skip_adapterremoval': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description=None,
    ),
    'skip_preseq': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description=None,
    ),
    'skip_deduplication': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description=None,
    ),
    'skip_damage_calculation': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description=None,
    ),
    'skip_qualimap': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description=None,
    ),
    'complexity_filter_poly_g': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Complexity filtering',
        description='Turn on running poly-G removal on FASTQ files. Will only be performed on 2 colour chemistry machine sequenced libraries.',
    ),
    'complexity_filter_poly_g_min': NextflowParameter(
        type=typing.Optional[int],
        default=10,
        section_title=None,
        description='Specify length of poly-g min for clipping to be performed.',
    ),
    'clip_forward_adaptor': NextflowParameter(
        type=typing.Optional[str],
        default='AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC',
        section_title='Read merging and adapter removal',
        description='Specify adapter sequence to be clipped off (forward strand).',
    ),
    'clip_reverse_adaptor': NextflowParameter(
        type=typing.Optional[str],
        default='AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTA',
        section_title=None,
        description='Specify adapter sequence to be clipped off (reverse strand).',
    ),
    'clip_adapters_list': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to AdapterRemoval adapter list file. Overrides `--clip_*_adaptor` parameters',
    ),
    'clip_readlength': NextflowParameter(
        type=typing.Optional[int],
        default=30,
        section_title=None,
        description='Specify read minimum length to be kept for downstream analysis.',
    ),
    'clip_min_read_quality': NextflowParameter(
        type=typing.Optional[int],
        default=20,
        section_title=None,
        description='Specify minimum base quality for trimming off bases.',
    ),
    'min_adap_overlap': NextflowParameter(
        type=typing.Optional[int],
        default=1,
        section_title=None,
        description='Specify minimum adapter overlap required for clipping.',
    ),
    'skip_collapse': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Skip of merging forward and reverse reads together and turns on paired-end alignment for downstream mapping. Only applicable for paired-end libraries.',
    ),
    'skip_trim': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Skip adapter and quality trimming.',
    ),
    'preserve5p': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Skip quality base trimming (n, score, window) of 5 prime end.',
    ),
    'mergedonly': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Only use merged reads downstream (un-merged reads and singletons are discarded).',
    ),
    'qualitymax': NextflowParameter(
        type=typing.Optional[int],
        default=41,
        section_title=None,
        description='Specify the maximum Phred score used in input FASTQ files',
    ),
    'run_post_ar_trimming': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on trimming of inline barcodes (i.e. internal barcodes after adapter removal)',
    ),
    'post_ar_trim_front': NextflowParameter(
        type=typing.Optional[int],
        default=7,
        section_title=None,
        description='Specify the number of bases to trim off the front of a merged read or R1',
    ),
    'post_ar_trim_tail': NextflowParameter(
        type=typing.Optional[int],
        default=7,
        section_title=None,
        description='Specify the number of bases to trim off the tail of of a merged read or R1',
    ),
    'post_ar_trim_front2': NextflowParameter(
        type=typing.Optional[int],
        default=7,
        section_title=None,
        description='Specify the number of bases to trim off the front of R2',
    ),
    'post_ar_trim_tail2': NextflowParameter(
        type=typing.Optional[int],
        default=7,
        section_title=None,
        description='Specify the number of bases to trim off the tail of R2',
    ),
    'mapper': NextflowParameter(
        type=typing.Optional[str],
        default='bwaaln',
        section_title='Read mapping to reference genome',
        description="Specify which mapper to use. Options: 'bwaaln', 'bwamem', 'circularmapper', 'bowtie2'.",
    ),
    'bwaalnn': NextflowParameter(
        type=typing.Optional[float],
        default=0.01,
        section_title=None,
        description='Specify the -n parameter for BWA aln, i.e. amount of allowed mismatches in the alignment.',
    ),
    'bwaalnk': NextflowParameter(
        type=typing.Optional[int],
        default=2,
        section_title=None,
        description='Specify the -k parameter for BWA aln, i.e. maximum edit distance allowed in a seed.',
    ),
    'bwaalnl': NextflowParameter(
        type=typing.Optional[int],
        default=1024,
        section_title=None,
        description='Specify the -l parameter for BWA aln i.e. the length of seeds to be used.',
    ),
    'bwaalno': NextflowParameter(
        type=typing.Optional[int],
        default=2,
        section_title=None,
        description='Specify the -o parameter for BWA aln i.e. the number of gaps allowed.',
    ),
    'circularextension': NextflowParameter(
        type=typing.Optional[int],
        default=500,
        section_title=None,
        description='Specify the number of bases to extend reference by (circularmapper only).',
    ),
    'circulartarget': NextflowParameter(
        type=typing.Optional[str],
        default='MT',
        section_title=None,
        description='Specify the FASTA header of the target chromosome to extend (circularmapper only).',
    ),
    'circularfilter': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on to remove reads that did not map to the circularised genome (circularmapper only).',
    ),
    'bt2_alignmode': NextflowParameter(
        type=typing.Optional[str],
        default='local',
        section_title=None,
        description="Specify the bowtie2 alignment mode. Options:  'local', 'end-to-end'.",
    ),
    'bt2_sensitivity': NextflowParameter(
        type=typing.Optional[str],
        default='sensitive',
        section_title=None,
        description="Specify the level of sensitivity for the bowtie2 alignment mode. Options: 'no-preset', 'very-fast', 'fast', 'sensitive', 'very-sensitive'.",
    ),
    'bt2n': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Specify the -N parameter for bowtie2 (mismatches in seed). This will override defaults from alignmode/sensitivity.',
    ),
    'bt2l': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Specify the -L parameter for bowtie2 (length of seed substrings). This will override defaults from alignmode/sensitivity.',
    ),
    'bt2_trim5': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify number of bases to trim off from 5' (left) end of read before alignment.",
    ),
    'bt2_trim3': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify number of bases to trim off from 3' (right) end of read before alignment.",
    ),
    'bt2_maxins': NextflowParameter(
        type=typing.Optional[int],
        default=500,
        section_title=None,
        description='Specify the maximum fragment length for Bowtie2 paired-end mapping mode only.',
    ),
    'hostremoval_input_fastq': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Removal of Host-Mapped Reads',
        description='Turn on per-library creation pre-Adapter Removal FASTQ files without reads that mapped to reference (e.g. for public upload of privacy sensitive non-host data)',
    ),
    'hostremoval_mode': NextflowParameter(
        type=typing.Optional[str],
        default='remove',
        section_title=None,
        description='Host removal mode. Remove mapped reads completely from FASTQ (remove) or just mask mapped reads sequence by N (replace).',
    ),
    'run_bam_filtering': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='BAM Filtering',
        description='Turn on filtering of mapping quality, read lengths, or unmapped reads of BAM files.',
    ),
    'bam_mapping_quality_threshold': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Minimum mapping quality for reads filter.',
    ),
    'bam_filter_minreadlength': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Specify minimum read length to be kept after mapping.',
    ),
    'bam_unmapped_type': NextflowParameter(
        type=typing.Optional[str],
        default='discard',
        section_title=None,
        description="Defines whether to discard all unmapped reads, keep only bam and/or keep only fastq format Options: 'discard', 'bam', 'fastq', 'both'.",
    ),
    'dedupper': NextflowParameter(
        type=typing.Optional[str],
        default='markduplicates',
        section_title='DeDuplication',
        description="Deduplication method to use. Options: 'markduplicates',  'dedup'.",
    ),
    'dedup_all_merged': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on treating all reads as merged reads.',
    ),
    'preseq_mode': NextflowParameter(
        type=typing.Optional[str],
        default='c_curve',
        section_title='Library Complexity Analysis',
        description='Specify which mode of preseq to run.',
    ),
    'preseq_step_size': NextflowParameter(
        type=typing.Optional[int],
        default=1000,
        section_title=None,
        description='Specify the step size of Preseq.',
    ),
    'preseq_maxextrap': NextflowParameter(
        type=typing.Optional[int],
        default=10000000000,
        section_title=None,
        description='Specify the maximum extrapolation (lc_extrap mode only)',
    ),
    'preseq_terms': NextflowParameter(
        type=typing.Optional[int],
        default=100,
        section_title=None,
        description='Specify the maximum number of terms for extrapolation (lc_extrap mode only)',
    ),
    'preseq_bootstrap': NextflowParameter(
        type=typing.Optional[int],
        default=100,
        section_title=None,
        description='Specify number of bootstraps to perform (lc_extrap mode only)',
    ),
    'preseq_cval': NextflowParameter(
        type=typing.Optional[float],
        default=0.95,
        section_title=None,
        description='Specify confidence interval level (lc_extrap mode only)',
    ),
    'damage_calculation_tool': NextflowParameter(
        type=typing.Optional[str],
        default='damageprofiler',
        section_title='(aDNA) Damage Analysis',
        description='Specify the tool to use for damage calculation.',
    ),
    'damageprofiler_length': NextflowParameter(
        type=typing.Optional[int],
        default=100,
        section_title=None,
        description='Specify length filter for DamageProfiler.',
    ),
    'damageprofiler_threshold': NextflowParameter(
        type=typing.Optional[int],
        default=15,
        section_title=None,
        description='Specify number of bases of each read to consider for DamageProfiler calculations.',
    ),
    'damageprofiler_yaxis': NextflowParameter(
        type=typing.Optional[float],
        default=0.3,
        section_title=None,
        description="Specify the maximum misincorporation frequency that should be displayed on the damage plot. Set to 0 to 'autoscale'.",
    ),
    'mapdamage_downsample': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Specify the maximum number of reads to consider for damage calculation. Defaults value is `0` (i.e. no downsampling is performed).',
    ),
    'mapdamage_yaxis': NextflowParameter(
        type=typing.Optional[float],
        default=0.3,
        section_title=None,
        description='Specify the maximum misincorporation frequency that should be displayed on the damage plot.',
    ),
    'run_pmdtools': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on PMDtools',
    ),
    'pmdtools_range': NextflowParameter(
        type=typing.Optional[int],
        default=10,
        section_title=None,
        description='Specify range of bases for PMDTools to scan for damage.',
    ),
    'pmdtools_threshold': NextflowParameter(
        type=typing.Optional[int],
        default=3,
        section_title=None,
        description='Specify PMDScore threshold for PMDTools.',
    ),
    'pmdtools_reference_mask': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify a bedfile to be used to mask the reference fasta prior to running pmdtools.',
    ),
    'pmdtools_max_reads': NextflowParameter(
        type=typing.Optional[int],
        default=10000,
        section_title=None,
        description='Specify the maximum number of reads to consider for metrics generation.',
    ),
    'pmdtools_platypus': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Append big list of base frequencies for platypus to output.',
    ),
    'run_mapdamage_rescaling': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on damage rescaling of BAM files using mapDamage2 to probabilistically remove damage.',
    ),
    'rescale_seqlength': NextflowParameter(
        type=typing.Optional[int],
        default=12,
        section_title=None,
        description='Length of read sequence to use from each side for rescaling. Can be overridden by --rescale_length_*p.',
    ),
    'rescale_length_5p': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Length of read for mapDamage2 to rescale from 5p end.  Only used if not 0 otherwise --rescale_seqlength used.',
    ),
    'rescale_length_3p': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Length of read for mapDamage2 to rescale from 3p end. Only used if not 0 otherwise --rescale_seqlength used..',
    ),
    'run_bedtools_coverage': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Feature Annotation Statistics',
        description='Turn on ability to calculate no. reads, depth and breadth coverage of features in reference.',
    ),
    'anno_file': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to GFF or BED file containing positions of features in reference file (--fasta). Path should be enclosed in quotes.',
    ),
    'anno_file_is_unsorted': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specify if the annotation file provided to --anno_file is not sorted in the same way as the reference fasta file.',
    ),
    'run_trim_bam': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='BAM Trimming',
        description='Turn on BAM trimming. Will only run on non-UDG or half-UDG libraries',
    ),
    'bamutils_clip_double_stranded_half_udg_left': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'left' end of read for double-stranded half-UDG libraries.",
    ),
    'bamutils_clip_double_stranded_half_udg_right': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'right' end of read for double-stranded half-UDG libraries.",
    ),
    'bamutils_clip_double_stranded_none_udg_left': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'left' end of read for double-stranded non-UDG libraries.",
    ),
    'bamutils_clip_double_stranded_none_udg_right': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'right' end of read for double-stranded non-UDG libraries.",
    ),
    'bamutils_clip_single_stranded_half_udg_left': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'left' end of read for single-stranded half-UDG libraries.",
    ),
    'bamutils_clip_single_stranded_half_udg_right': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'right' end of read for single-stranded half-UDG libraries.",
    ),
    'bamutils_clip_single_stranded_none_udg_left': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'left' end of read for single-stranded non-UDG libraries.",
    ),
    'bamutils_clip_single_stranded_none_udg_right': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description="Specify the number of bases to clip off reads from 'right' end of read for single-stranded non-UDG libraries.",
    ),
    'bamutils_softclip': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on using softclip instead of hard masking.',
    ),
    'run_genotyping': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Genotyping',
        description='Turn on genotyping of BAM files.',
    ),
    'genotyping_tool': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Specify which genotyper to use either GATK UnifiedGenotyper, GATK HaplotypeCaller, Freebayes, or pileupCaller. Options: 'ug', 'hc', 'freebayes', 'pileupcaller', 'angsd'.",
    ),
    'genotyping_source': NextflowParameter(
        type=typing.Optional[str],
        default='raw',
        section_title=None,
        description="Specify which input BAM to use for genotyping. Options: 'raw', 'trimmed', 'pmd' or 'rescaled'.",
    ),
    'gatk_call_conf': NextflowParameter(
        type=typing.Optional[int],
        default=30,
        section_title=None,
        description='Specify GATK phred-scaled confidence threshold.',
    ),
    'gatk_ploidy': NextflowParameter(
        type=typing.Optional[int],
        default=2,
        section_title=None,
        description='Specify GATK organism ploidy.',
    ),
    'gatk_downsample': NextflowParameter(
        type=typing.Optional[int],
        default=250,
        section_title=None,
        description='Maximum depth coverage allowed for genotyping before down-sampling is turned on.',
    ),
    'gatk_dbsnp': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify VCF file for SNP annotation of output VCF files. Optional. Gzip not accepted.',
    ),
    'gatk_hc_out_mode': NextflowParameter(
        type=typing.Optional[str],
        default='EMIT_VARIANTS_ONLY',
        section_title=None,
        description="Specify GATK output mode. Options: 'EMIT_VARIANTS_ONLY', 'EMIT_ALL_CONFIDENT_SITES', 'EMIT_ALL_ACTIVE_SITES'.",
    ),
    'gatk_hc_emitrefconf': NextflowParameter(
        type=typing.Optional[str],
        default='GVCF',
        section_title=None,
        description="Specify HaplotypeCaller mode for emitting reference confidence calls . Options: 'NONE', 'BP_RESOLUTION', 'GVCF'.",
    ),
    'gatk_ug_out_mode': NextflowParameter(
        type=typing.Optional[str],
        default='EMIT_VARIANTS_ONLY',
        section_title=None,
        description="Specify GATK output mode. Options: 'EMIT_VARIANTS_ONLY', 'EMIT_ALL_CONFIDENT_SITES', 'EMIT_ALL_SITES'.",
    ),
    'gatk_ug_genotype_model': NextflowParameter(
        type=typing.Optional[str],
        default='SNP',
        section_title=None,
        description="Specify UnifiedGenotyper likelihood model. Options: 'SNP', 'INDEL', 'BOTH', 'GENERALPLOIDYSNP', 'GENERALPLOIDYINDEL'.",
    ),
    'gatk_ug_keep_realign_bam': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specify to keep the BAM output of re-alignment around variants from GATK UnifiedGenotyper.',
    ),
    'gatk_ug_defaultbasequalities': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Supply a default base quality if a read is missing a base quality score. Setting to -1 turns this off.',
    ),
    'freebayes_C': NextflowParameter(
        type=typing.Optional[int],
        default=1,
        section_title=None,
        description='Specify minimum required supporting observations to consider a variant.',
    ),
    'freebayes_g': NextflowParameter(
        type=typing.Optional[int],
        default=0,
        section_title=None,
        description='Specify to skip over regions of high depth by discarding alignments overlapping positions where total read depth is greater than specified in --freebayes_C.',
    ),
    'freebayes_p': NextflowParameter(
        type=typing.Optional[int],
        default=2,
        section_title=None,
        description='Specify ploidy of sample in FreeBayes.',
    ),
    'pileupcaller_bedfile': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify path to SNP panel in bed format for pileupCaller.',
    ),
    'pileupcaller_snpfile': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify path to SNP panel in EIGENSTRAT format for pileupCaller.',
    ),
    'pileupcaller_method': NextflowParameter(
        type=typing.Optional[str],
        default='randomHaploid',
        section_title=None,
        description="Specify calling method to use. Options: 'randomHaploid', 'randomDiploid', 'majorityCall'.",
    ),
    'pileupcaller_transitions_mode': NextflowParameter(
        type=typing.Optional[str],
        default='AllSites',
        section_title=None,
        description="Specify the calling mode for transitions. Options: 'AllSites', 'TransitionsMissing', 'SkipTransitions'.",
    ),
    'pileupcaller_min_map_quality': NextflowParameter(
        type=typing.Optional[int],
        default=30,
        section_title=None,
        description='The minimum mapping quality to be used for genotyping.',
    ),
    'pileupcaller_min_base_quality': NextflowParameter(
        type=typing.Optional[int],
        default=30,
        section_title=None,
        description='The minimum base quality to be used for genotyping.',
    ),
    'angsd_glmodel': NextflowParameter(
        type=typing.Optional[str],
        default='samtools',
        section_title=None,
        description="Specify which ANGSD genotyping likelihood model to use. Options: 'samtools', 'gatk', 'soapsnp', 'syk'.",
    ),
    'angsd_glformat': NextflowParameter(
        type=typing.Optional[str],
        default='binary',
        section_title=None,
        description="Specify which output type to output ANGSD genotyping likelihood results: Options: 'text', 'binary', 'binary_three', 'beagle'.",
    ),
    'angsd_createfasta': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on creation of FASTA from ANGSD genotyping likelihood.',
    ),
    'angsd_fastamethod': NextflowParameter(
        type=typing.Optional[str],
        default='random',
        section_title=None,
        description="Specify which genotype type of 'base calling' to use for ANGSD FASTA generation. Options: 'random', 'common'.",
    ),
    'run_bcftools_stats': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Turn on bcftools stats generation for VCF based variant calling statistics',
    ),
    'run_vcf2genome': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Consensus Sequence Generation',
        description='Turns on ability to create a consensus sequence FASTA file based on a UnifiedGenotyper VCF file and the original reference (only considers SNPs).',
    ),
    'vcf2genome_outfile': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify the name of the output FASTA file containing the consensus sequence.',
    ),
    'vcf2genome_header': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify the header name of the consensus sequence entry within the FASTA file.',
    ),
    'vcf2genome_minc': NextflowParameter(
        type=typing.Optional[int],
        default=5,
        section_title=None,
        description='Minimum depth coverage required for a call to be included (else N will be called).',
    ),
    'vcf2genome_minq': NextflowParameter(
        type=typing.Optional[int],
        default=30,
        section_title=None,
        description='Minimum genotyping quality of a call to be called. Else N will be called.',
    ),
    'vcf2genome_minfreq': NextflowParameter(
        type=typing.Optional[float],
        default=0.8,
        section_title=None,
        description='Minimum fraction of reads supporting a call to be included. Else N will be called.',
    ),
    'run_multivcfanalyzer': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='SNP Table Generation',
        description='Turn on MultiVCFAnalyzer. Note: This currently only supports diploid GATK UnifiedGenotyper input.',
    ),
    'write_allele_frequencies': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on writing write allele frequencies in the SNP table.',
    ),
    'min_genotype_quality': NextflowParameter(
        type=typing.Optional[int],
        default=30,
        section_title=None,
        description='Specify the minimum genotyping quality threshold for a SNP to be called.',
    ),
    'min_base_coverage': NextflowParameter(
        type=typing.Optional[int],
        default=5,
        section_title=None,
        description='Specify the minimum number of reads a position needs to be covered to be considered for base calling.',
    ),
    'min_allele_freq_hom': NextflowParameter(
        type=typing.Optional[float],
        default=0.9,
        section_title=None,
        description="Specify the minimum allele frequency that a base requires to be considered a 'homozygous' call.",
    ),
    'min_allele_freq_het': NextflowParameter(
        type=typing.Optional[float],
        default=0.9,
        section_title=None,
        description="Specify the minimum allele frequency that a base requires to be considered a 'heterozygous' call.",
    ),
    'additional_vcf_files': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify paths to additional pre-made VCF files to be included in the SNP table generation. Use wildcard(s) for multiple files.',
    ),
    'reference_gff_annotations': NextflowParameter(
        type=typing.Optional[str],
        default='NA',
        section_title=None,
        description="Specify path to the reference genome annotations in '.gff' format. Optional.",
    ),
    'reference_gff_exclude': NextflowParameter(
        type=typing.Optional[str],
        default='NA',
        section_title=None,
        description="Specify path to the positions to be excluded in '.gff' format. Optional.",
    ),
    'snp_eff_results': NextflowParameter(
        type=typing.Optional[str],
        default='NA',
        section_title=None,
        description="Specify path to the output file from SNP effect analysis in '.txt' format. Optional.",
    ),
    'run_mtnucratio': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Mitochondrial to Nuclear Ratio',
        description='Turn on mitochondrial to nuclear ratio calculation.',
    ),
    'mtnucratio_header': NextflowParameter(
        type=typing.Optional[str],
        default='MT',
        section_title=None,
        description='Specify the name of the reference FASTA entry corresponding to the mitochondrial genome (up to the first space).',
    ),
    'run_sexdeterrmine': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Human Sex Determination',
        description='Turn on sex determination for human reference genomes. This will run on single- and double-stranded variants of a library separately.',
    ),
    'sexdeterrmine_bedfile': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify path to SNP panel in bed format for error bar calculation. Optional (see documentation).',
    ),
    'run_nuclear_contamination': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Nuclear Contamination for Human DNA',
        description='Turn on nuclear contamination estimation for human reference genomes.',
    ),
    'contamination_chrom_name': NextflowParameter(
        type=typing.Optional[str],
        default='X',
        section_title=None,
        description="The name of the X chromosome in your bam/FASTA header. 'X' for hs37d5, 'chrX' for HG19.",
    ),
    'metagenomic_complexity_filter': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Metagenomic Screening',
        description='Turn on removal of low-sequence complexity reads for metagenomic screening with bbduk',
    ),
    'metagenomic_complexity_entropy': NextflowParameter(
        type=typing.Optional[float],
        default=0.3,
        section_title=None,
        description='Specify the entropy threshold that under which a sequencing read will be complexity filtered out. This should be between 0-1.',
    ),
    'run_metagenomic_screening': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on metagenomic screening module for reference-unmapped reads.',
    ),
    'metagenomic_tool': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Specify which classifier to use. Options: 'malt', 'kraken'.",
    ),
    'database': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Specify path to classifier database directory. For Kraken2 this can also be a `.tar.gz` of the directory.',
    ),
    'metagenomic_min_support_reads': NextflowParameter(
        type=typing.Optional[int],
        default=1,
        section_title=None,
        description="Specify a minimum number of reads a taxon of sample total is required to have to be retained. Not compatible with --malt_min_support_mode 'percent'.",
    ),
    'percent_identity': NextflowParameter(
        type=typing.Optional[int],
        default=85,
        section_title=None,
        description='Percent identity value threshold for MALT.',
    ),
    'malt_mode': NextflowParameter(
        type=typing.Optional[str],
        default='BlastN',
        section_title=None,
        description="Specify which alignment mode to use for MALT. Options: 'Unknown', 'BlastN', 'BlastP', 'BlastX', 'Classifier'.",
    ),
    'malt_alignment_mode': NextflowParameter(
        type=typing.Optional[str],
        default='SemiGlobal',
        section_title=None,
        description="Specify alignment method for MALT. Options: 'Local', 'SemiGlobal'.",
    ),
    'malt_top_percent': NextflowParameter(
        type=typing.Optional[int],
        default=1,
        section_title=None,
        description='Specify the percent for LCA algorithm for MALT (see MEGAN6 CE manual).',
    ),
    'malt_min_support_mode': NextflowParameter(
        type=typing.Optional[str],
        default='percent',
        section_title=None,
        description="Specify whether to use percent or raw number of reads for minimum support required for taxon to be retained for MALT. Options: 'percent', 'reads'.",
    ),
    'malt_min_support_percent': NextflowParameter(
        type=typing.Optional[float],
        default=0.01,
        section_title=None,
        description='Specify the minimum percentage of reads a taxon of sample total is required to have to be retained for MALT.',
    ),
    'malt_max_queries': NextflowParameter(
        type=typing.Optional[int],
        default=100,
        section_title=None,
        description='Specify the maximum number of queries a read can have for MALT.',
    ),
    'malt_memory_mode': NextflowParameter(
        type=typing.Optional[str],
        default='load',
        section_title=None,
        description="Specify the memory load method. Do not use 'map' with GPFS file systems for MALT as can be very slow. Options: 'load', 'page', 'map'.",
    ),
    'malt_sam_output': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specify to also produce SAM alignment files. Note this includes both aligned and unaligned reads, and are gzipped. Note this will result in very large file sizes.',
    ),
    'run_maltextract': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title='Metagenomic Authentication',
        description='Turn on MaltExtract for MALT aDNA characteristics authentication.',
    ),
    'maltextract_taxon_list': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to a text file with taxa of interest (one taxon per row, NCBI taxonomy name format)',
    ),
    'maltextract_ncbifiles': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to directory containing containing NCBI resource files (ncbi.tre and ncbi.map; available: https://github.com/rhuebler/HOPS/)',
    ),
    'maltextract_filter': NextflowParameter(
        type=typing.Optional[str],
        default='def_anc',
        section_title=None,
        description="Specify which MaltExtract filter to use. Options: 'def_anc', 'ancient', 'default', 'crawl', 'scan', 'srna', 'assignment'.",
    ),
    'maltextract_toppercent': NextflowParameter(
        type=typing.Optional[float],
        default=0.01,
        section_title=None,
        description='Specify percent of top alignments to use.',
    ),
    'maltextract_destackingoff': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn off destacking.',
    ),
    'maltextract_downsamplingoff': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn off downsampling.',
    ),
    'maltextract_duplicateremovaloff': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn off duplicate removal.',
    ),
    'maltextract_matches': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on exporting alignments of hits in BLAST format.',
    ),
    'maltextract_megansummary': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on export of MEGAN summary files.',
    ),
    'maltextract_percentidentity': NextflowParameter(
        type=typing.Optional[float],
        default=85,
        section_title=None,
        description='Minimum percent identity alignments are required to have to be reported. Recommended to set same as MALT parameter.',
    ),
    'maltextract_topalignment': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Turn on using top alignments per read after filtering.',
    ),
}

